### PR TITLE
Refactor and optimize model evaluation 

### LINKED ISSUE #202: [FEA] Optimize `torch_gather_like` operator used in top-k extraction
# 🚀 Feature request

## Motivation
- Create a vectorized version of  torch_gather_like function implemented [here](https://github.com/NVIDIA-Merlin/models/blob/c0df6db16180d696400e7fab9747e907eb394cf4/merlin/models/tf/utils/tf_utils.py#L108). This function is used to gather top-k indices from each row of one-hot vector of labels. 

## Your contribution

Maybe later check if we could come up with a more efficient implementation of `gather_torch_like` which does not using `TensorArray`, but rather uses only `tf.gather_nd()` with a 2D index (rows, cols), like the one I implemented here [here](https://github.com/NVIDIA-Merlin/NVTabular/blob/dc9c04d126b8ad842a32a3618e8f4495afbe0678/nvtabular/loader/tensorflow.py#L404). 
Not a priority for now, but maybe it might improve the metrics performance

_Originally posted by @gabrielspmoreira in https://github.com/NVIDIA-Merlin/models/pull/189#discussion_r813052963_

### LINKED ISSUE #210: [FEA] Computing Training/Eval metrics after each n steps
# 🚀 Feature request

Create a custom Callback for computing metrics after each n-steps instead of computing them for each batch.  This callback can be then provided to the `fit` method.

## Motivation

- Evaluating the RetrievalModels trained with negative sampling strategies requires comparing the positive scores against a set of implicit negative ones using ranking metrics. One required step to compute those metrics is the extraction of the top-k negatives.  This process can be time-consuming (when the catalog of items get large) and the goal of this feature is to gather positive scores and labels for n steps before calling the top-k extraction and ranking metrics calculation.

### LINKED ISSUE #225: [BUG] Pre-sorting of predictions for metrics optimization causes incorrect metric calculation
Refactoring metrics pre-sorting computation to avoid incorrect calculation because pre-sorting and extracting top-k changes the number of relevant count for each row, which is not correct.

### PR DESCRIPTION
**Pull request by gabrielspmoreira**
This PR refactors model evaluation and also optimizes eval performance:

- [x] Renaming `pre_metrics` (e.g. BruteForceTopK) to `pre_eval_topk` (for both `ItemRetrievalTask` and `NextItemPredictionTask`), to make it clear it is only called during evaluation/prediction, and not when evaluating during training
- [x] Converting `tf_utils.gather_torch_like()` method into a vectorized version to retrieve the top-k labels. Closes #202 
- [x] Created the LabelToOneHot transformation block to centralized that logic, which was done in many places. Now it is expected that when ranking metrics are used the labels must be necessarily one-hot encoded, as that is required by `extract_topk`.
- [x] Encapsulated internal properties with setter methods for better maintainability (e.g. `self.loss_block.retrieval_scorer.cache_query` and self.loss_block.pre_eval_topk)
- [x] Refactored metrics pre-sorting computation to avoid incorrect calculation (because pre-sorting and extracting top-k changes the number of relevant count for each row, which is not correct). Closes #225 
- [x] Refactored ranking metrics to have a single instance for each k (rather one metric with many `top_ks`), which simplifies naming and collecting results 
- [ ] Refactored ranking metrics to accumulate only the sum of batch metrics and the batches count, to aggregate results the same way default Keras metrics does (simple average over the epoch by dividing the sum by the count) **(in progress)**
- [ ] Making train metrics to be computed each `train_metrics_steps`. If `train_metrics_steps=0` no metrics are computed during training. Closes #210  **(In progress)**

### CODE REVIEW

File: merlin/models/tf/blocks/core/transformations.py [hunk 1]
```diff
@@ -400,3 +401,23 @@ def call(self, inputs, training=False, **kwargs) -> tf.Tensor:
...
+
+
+@Block.registry.register_with_multiple_names("label_to_onehot")
+@tf.keras.utils.register_keras_serializable(package="merlin_models")
+class LabelToOneHot(Block):
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    def compute_output_shape(self, input_shape):
+        return input_shape
```
Comment (marcromeyn) [inline merlin/models/tf/blocks/core/transformations.py:413]: Doesn’t the shape change here?

File: merlin/models/tf/core.py [hunk 2]
```diff
@@ -1788,14 +1795,22 @@ def __init__(
...
         self._task_name = task_name
         self.pre = pre
-        self.pre_metrics = pre_metrics
+        self.pre_eval_topk = pre_eval_topk
+
+        if metrics is not None:
+            if isinstance(metrics, SequenceCollection):
+                metrics = list(metrics)
+            else:
+                metrics = [metrics]
```
Comment (marcromeyn) [inline merlin/models/tf/core.py:1804]: Seems like this allows you to pass in a single metric, if that’s the case we should update the type-hint in the class.

File: merlin/models/tf/core.py [hunk 3]
```diff
@@ -1788,14 +1795,22 @@ def __init__(
...
 
         create_metrics = self._create_metrics
         self.eval_metrics = create_metrics(metrics) if metrics else []
         self.prediction_metrics = create_metrics(prediction_metrics) if prediction_metrics else []
         self.label_metrics = create_metrics(label_metrics) if label_metrics else []
         self.loss_metrics = create_metrics(loss_metrics) if loss_metrics else []
-        self.compute_train_metrics = compute_train_metrics
+
+    def set_pre_eval_topk(self, block):
```
Comment (marcromeyn) [inline merlin/models/tf/core.py:1812]: Why do we want a setter for this? If it’s for safety we might want to make the variable private and expose it as a property.

File: merlin/models/tf/core.py [hunk 4]
```diff
@@ -2274,6 +2367,33 @@ def get_config(self):
         return {"block": tf.keras.utils.serialize_keras_object(self.block)}
 
 
+class MetricsComputeCallback(tf.keras.callbacks.Callback):
+    def __init__(self, train_metrics_steps=1, **kwargs):
```
Comment (marcromeyn) [inline merlin/models/tf/core.py:2371]: Maybe we should log how often we update the metrics here.

File: merlin/models/tf/core.py [hunk 5]
```diff
@@ -2491,6 +2627,12 @@ def fit(
 
             x = Dataset(x, batch_size=batch_size, **kwargs)
 
+        if callbacks is None:
+            callbacks = []
+
+        # Adding a callback to control metrics computation
+        callbacks.append(MetricsComputeCallback(train_metrics_steps))
```
Comment (marcromeyn) [inline merlin/models/tf/core.py:2634]: We probably should check whether `MetricsComputeCallback` is already part of the callbacks.

File: merlin/models/tf/core.py [hunk 6]
```diff
@@ -2683,8 +2823,8 @@ def load_topk_evaluation(self, data: merlin.io.Dataset, k: int, **kwargs):
             self.retrieval_block.item_block(), data=data, k=k - 1, context=self.context, **kwargs
         )
         # set cache_query to True in the ItemRetrievalScorer
-        self.loss_block.retrieval_scorer.cache_query = True
-        self.loss_block.pre_metrics = topk_index
+        self.loss_block.set_retrieval_cache_query(True)
```
Comment (marcromeyn) [inline merlin/models/tf/core.py:2826]: I think we should we check whether these functions are defined in the loss_block first. [This](https://github.com/NVIDIA-Merlin/models/blob/65e78feaac37fcc831de1fc7d043e4eb1f4751f9/merlin/models/tf/utils/mixins.py#L98) is the type, so it could not be there.

File: merlin/models/tf/blocks/core/transformations.py (hunk 1 above)
Comment (gabrielspmoreira) [reply merlin/models/tf/blocks/core/transformations.py:413]: I debugged it and input_shape comes with the correct dims `TensorShape(batch size, # items)`. In this transform, we only apply one-hot encoding to the `labels`, which in fact have there shape changed to be equal to `predictions`.

File: merlin/models/tf/core.py (hunk 2 above)
Comment (gabrielspmoreira) [reply merlin/models/tf/core.py:1804]: Created a new `MetricOrMetrics = Union[Sequence[MetricOrMetricClass], MetricOrMetricClass]` that is now used in tasks metrics arguments

File: merlin/models/tf/core.py (hunk 3 above)
Comment (gabrielspmoreira) [reply merlin/models/tf/core.py:1812]: Yes, it is to protect the internal variable. I did like you suggested

File: merlin/models/tf/core.py (hunk 6 above)
Comment (gabrielspmoreira) [reply merlin/models/tf/core.py:2826]: Now I am checking if `self.loss_block` is defined and if it is an instance of `ItemRetrievalTask`

File: merlin/models/tf/core.py (hunk 4 above)
Comment (gabrielspmoreira) [reply merlin/models/tf/core.py:2385]: When we fit() with keras, we can observe that metrics are only updated each N steps, keeping the last result for the other steps. That way we can verify that the metrics are really being computed each N steps.
